AnnService/inc/Core/Common/BKTree.h

```diff
@@ -106,23 +107,36 @@ namespace SPTAG
         class BKTree
         {
         public:
-            BKTree(): m_iTreeNumber(1), m_iBKTKmeansK(32), m_iBKTLeafSize(8), m_iSamples(1000) {}
+            BKTree(): m_iTreeNumber(1), m_iBKTKmeansK(32), m_iBKTLeafSize(8), m_iSamples(1000), m_lock(new std::shared_timed_mutex) {}
             
             BKTree(BKTree& other): m_iTreeNumber(other.m_iTreeNumber), 
                                    m_iBKTKmeansK(other.m_iBKTKmeansK), 
                                    m_iBKTLeafSize(other.m_iBKTLeafSize),
-                                   m_iSamples(other.m_iSamples) {}
+                                   m_iSamples(other.m_iSamples),
+                                   m_lock(new std::shared_timed_mutex) {}
             ~BKTree() {}
 
             inline const BKTNode& operator[](SizeType index) const { return m_pTreeRoots[index]; }
             inline BKTNode& operator[](SizeType index) { return m_pTreeRoots[index]; }
 
             inline SizeType size() const { return (SizeType)m_pTreeRoots.size(); }
+            
+            inline SizeType sizePerTree() const {
+                std::shared_lock<std::shared_timed_mutex> lock(*m_lock);
+                return (SizeType)m_pTreeRoots.size() - m_pTreeStart.back(); 
+            }
 
             inline const std::unordered_map<SizeType, SizeType>& GetSampleMap() const { return m_pSampleCenterMap; }
 
+            void swap(BKTree& p_tree)
+            {
+                std::unique_lock<std::shared_timed_mutex> lock(*m_lock);
```

Comment:
should p_tree.m_lock be also locked?

Reply:
m_lock is used to avoid doing search when trees are being updated. Here p_tree is the new version of trees which will not be used in any search process. It does not need to be locked.